dbms/src/Interpreters/JoinV2/HashJoinProbe.cpp

```diff
@@ -0,0 +1,667 @@
+// limitations under the License.
+
+#include <Columns/ColumnUtils.h>
+#include <Common/Stopwatch.h>
+#include <Interpreters/JoinUtils.h>
+#include <Interpreters/JoinV2/HashJoinProbe.h>
+#include <Interpreters/NullableUtils.h>
+
+#ifdef TIFLASH_ENABLE_AVX_SUPPORT
+ASSERT_USE_AVX2_COMPILE_FLAG
+#endif
+
+namespace DB
+{
+
+using enum ASTTableJoin::Kind;
+
+bool JoinProbeContext::isCurrentProbeFinished() const
+{
+    return start_row_idx >= rows && prefetch_active_states == 0;
+}
+
+void JoinProbeContext::resetBlock(Block & block_)
+{
+    block = block_;
+    orignal_block = block_;
+    rows = block.rows();
+    start_row_idx = 0;
+    current_probe_row_ptr = nullptr;
+    prefetch_active_states = 0;
+
+    is_prepared = false;
+    materialized_columns.clear();
+    key_columns.clear();
+    null_map = nullptr;
+    null_map_holder = nullptr;
+    current_row_is_matched = false;
+}
+
+void JoinProbeContext::prepareForHashProbe(
+    HashJoinKeyMethod method,
+    ASTTableJoin::Kind kind,
+    const Names & key_names,
+    const String & filter_column,
+    const NameSet & probe_output_name_set,
+    const TiDB::TiDBCollators & collators,
+    const HashJoinRowLayout & row_layout)
+{
+    if (is_prepared)
+        return;
+
+    key_columns = extractAndMaterializeKeyColumns(block, materialized_columns, key_names);
+    /// Some useless columns maybe key columns so they must be removed after extracting key columns.
+    for (size_t pos = 0; pos < block.columns();)
+    {
+        if (!probe_output_name_set.contains(block.getByPosition(pos).name))
+            block.erase(pos);
+        else
+            ++pos;
+    }
+
+    /// Keys with NULL value in any column won't join to anything.
+    extractNestedColumnsAndNullMap(key_columns, null_map_holder, null_map);
+    /// reuse null_map to record the filtered rows, the rows contains NULL or does not
+    /// match the join filter won't join to anything
+    recordFilteredRows(block, filter_column, null_map_holder, null_map);
+
+    if unlikely (!key_getter)
+        key_getter = createHashJoinKeyGetter(method, collators);
+
+    resetHashJoinKeyGetter(method, key_getter, key_columns, row_layout);
+
+    /** If you use FULL or RIGHT JOIN, then the columns from the "left" table must be materialized.
+     * Because if they are constants, then in the "not joined" rows, they may have different values
+     *  - default values, which can differ from the values of these constants.
+     */
+    if (getFullness(kind))
+    {
+        size_t existing_columns = block.columns();
+        for (size_t i = 0; i < existing_columns; ++i)
+        {
+            auto & col = block.getByPosition(i).column;
+
+            if (ColumnPtr converted = col->convertToFullColumnIfConst())
+                col = converted;
+
+            /// convert left columns (except keys) to Nullable
+            if (std::end(key_names) == std::find(key_names.begin(), key_names.end(), block.getByPosition(i).name))
```

Comment:
why left keys column don't need to be converted to nullable?

Reply:
Fixed